File: benchmarks/runtime/programs/go/constant-time.go
// constant-time — the equivalent-semantics Go peer for the tuonelang
// constant-time workload (ADR-0020), and the parity target it answers.
//
// Go's answer to this problem is `crypto/subtle.ConstantTimeCompare`, so that
// is what this peer uses — the same convention every other workload here
// follows, where the Go side is the standard library's own solution rather
// than a hand-transliteration of the tuonelang source (sha256-hash uses
// crypto/sha256, json-parse uses encoding/json). The comparison being drawn is
// "what does this cost in tuonelang versus what it costs in Go", and for a
// language with a standard-library answer that means using it.
//
// The naive early-returning comparison is written out, because Go's standard
// library deliberately does not provide the vulnerable form.
//
// Same 32-byte tags, same alternating best-case/worst-case inputs, same 1000
// rounds, same exit byte 32.
package main

import (
	"crypto/subtle"
	"os"
)

// The early-returning comparison: the vulnerability, for comparison.
func naiveBytesEq(a, b []byte) int {
	for i := range a {
		if a[i] != b[i] {
			return 0
		}
	}
	return 1
}

func makeTag(seed int) []byte {
	out := make([]byte, 32)
	for i := range out {
		out[i] = byte((seed + i*7) & 255)
	}
	return out
}

func main() {
	reference := makeTag(11)
	earlyMismatch := makeTag(12)
	equal := makeTag(11)

	agreements := 0
	for round := 0; round < 500000; round++ {
		var sameCT, sameNaive int
		if round&1 == 0 {
			sameCT = subtle.ConstantTimeCompare(reference, earlyMismatch)
			sameNaive = naiveBytesEq(reference, earlyMismatch)
		} else {
			sameCT = subtle.ConstantTimeCompare(reference, equal)
			sameNaive = naiveBytesEq(reference, equal)
		}
		if sameCT == sameNaive {
			agreements++
		}
	}
	os.Exit(agreements % 256)
}

File: benchmarks/runtime/programs/c/constant-time.c
/* constant-time — the equivalent-semantics C peer for the tuonelang
 * constant-time workload (ADR-0020). The identical computation: the same
 * 32-byte tags, the same branchless and early-returning comparisons over the
 * same alternating best-case/worst-case inputs, the same 1000 rounds.
 *
 * The masking idioms are written the same way as the tuonelang side rather
 * than the way C would conventionally write them. In particular the mask uses
 * the sign-smearing shift `(bit << 63) >> 63` instead of the conventional
 * `0 - bit`: tuonelang's `-` traps on i64::MIN, so its `std::ct` cannot use
 * the conventional spelling, and a peer that used the cheaper idiom would be
 * measuring a different program. Equivalent semantics means the same
 * algorithm, not merely the same answer.
 *
 * Signed shifts are used deliberately so the arithmetic right shift smears the
 * sign bit, matching tuonelang's Int (a signed 64-bit type). 500,000
 * agreements reduced mod 256 gives the exit byte 32. */
#include <stdint.h>

static int64_t ct_or_reduce(int64_t x) {
    int64_t a = x | (x >> 32);
    int64_t b = a | (a >> 16);
    int64_t c = b | (b >> 8);
    int64_t d = c | (c >> 4);
    int64_t e = d | (d >> 2);
    int64_t f = e | (e >> 1);
    return f & 1;
}

static int64_t ct_is_zero(int64_t x) { return ct_or_reduce(x) ^ 1; }

/* The branchless comparison: constant work for a given length. */
static int64_t ct_bytes_eq(const int64_t *a, const int64_t *b, int64_t n) {
    int64_t diff = 0;
    for (int64_t i = 0; i < n; i++) {
        diff |= a[i] ^ b[i];
    }
    return ct_is_zero(diff);
}

/* The early-returning comparison: the vulnerability, for comparison. */
static int64_t naive_bytes_eq(const int64_t *a, const int64_t *b, int64_t n) {
    for (int64_t i = 0; i < n; i++) {
        if (a[i] != b[i]) {
            return 0;
        }
    }
    return 1;
}

static void make_tag(int64_t seed, int64_t *out) {
    for (int64_t i = 0; i < 32; i++) {
        out[i] = (seed + i * 7) & 255;
    }
}

int main(void) {
    int64_t reference[32], early_mismatch[32], equal[32];
    make_tag(11, reference);
    make_tag(12, early_mismatch);
    make_tag(11, equal);

    int64_t agreements = 0;
    for (int64_t round = 0; round < 500000; round++) {
        int64_t same_ct, same_naive;
        if ((round & 1) == 0) {
            same_ct = ct_bytes_eq(reference, early_mismatch, 32);
            same_naive = naive_bytes_eq(reference, early_mismatch, 32);
        } else {
            same_ct = ct_bytes_eq(reference, equal, 32);
            same_naive = naive_bytes_eq(reference, equal, 32);
        }
        if (same_ct == same_naive) {
            agreements++;
        }
    }
    return (int)(agreements % 256);
}
